File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/messages/MessageReactionProtocol.kt
package dk.betterlectio.android.feature.messages

import java.net.URI
import java.text.Normalizer
import java.time.format.DateTimeFormatter
import java.util.Base64
import java.util.Locale
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import org.jsoup.Jsoup
import org.jsoup.nodes.Element

object MessageReactionProtocol {
    const val DOWNLOAD_URL = "https://betterlectio.dk/download"
    const val FRAGMENT_PREFIX = "blr1."
    private const val MAX_URL_LENGTH = 2_048
    private val json = Json { ignoreUnknownKeys = false }
    private val danishLocale = Locale.forLanguageTag("da-DK")
    private val isoSeconds = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss")

    sealed interface Envelope {
        val target: MessageLocator

        data class Set(
            val emoji: MessageReactionEmoji,
            override val target: MessageLocator,
        ) : Envelope

        data class Clear(
            override val target: MessageLocator,
        ) : Envelope
    }

    data class RawMessage(
        val entry: ThreadEntry,
        val rawContentHtml: String,
        val editPostbackTarget: String = "",
    )

    data class Carrier(
        val envelope: Envelope,
        val actor: MessageReactionParticipant,
        val editPostbackTarget: String,
        val index: Int,
    )

    data class ResolvedThread(
        val entries: List<ThreadEntry>,
        val ownCarriersByTarget: Map<MessageLocator, Carrier>,
        val hiddenCarrierCount: Int,
    )

    fun encode(envelope: Envelope): String {
        val payload = buildJsonObject {
            put("v", 1)
            when (envelope) {
                is Envelope.Set -> {
                    put("op", "set")
                    put("emoji", envelope.emoji.glyph)
                }
                is Envelope.Clear -> {
                    put("op", "clear")
                    put("emoji", JsonNull)
                }
            }
            put("target", buildJsonObject {
                put("senderKey", envelope.target.senderKey)
                put("sentAt", envelope.target.sentAt)
                put("occurrence", envelope.target.occurrence)
            })
        }.toString()
        return Base64.getUrlEncoder().withoutPadding().encodeToString(payload.toByteArray(Charsets.UTF_8))
    }

    fun decode(encoded: String): Envelope? = runCatching {
        if (!encoded.matches(Regex("^[A-Za-z0-9_-]+$"))) return null
        val root = json.parseToJsonElement(
            Base64.getUrlDecoder().decode(encoded).toString(Charsets.UTF_8),
        ).jsonObject
        if (root["v"]?.jsonPrimitive?.intOrNull != 1) return null
        val targetObject = root["target"]?.jsonObject ?: return null
        val target = MessageLocator(
            senderKey = targetObject["senderKey"]?.jsonPrimitive?.content ?: return null,
            sentAt = targetObject["sentAt"]?.jsonPrimitive?.content ?: return null,
            occurrence = targetObject["occurrence"]?.jsonPrimitive?.intOrNull ?: return null,
        )
        if (!isValidLocator(target)) return null
        when (root["op"]?.jsonPrimitive?.content) {
            "set" -> Envelope.Set(
                emoji = MessageReactionEmoji.fromGlyph(root["emoji"]?.jsonPrimitive?.content) ?: return null,
                target = target,
            )
            "clear" -> {
                if (root["emoji"] !is JsonNull) return null
                Envelope.Clear(target)
            }
            else -> null
        }
    }.getOrNull()

    fun carrierUrl(envelope: Envelope): String = "$DOWNLOAD_URL#$FRAGMENT_PREFIX${encode(envelope)}"

    fun carrierBody(envelope: Envelope, showSignature: Boolean): String {
        val sentence = when (envelope) {
            is Envelope.Set -> "Reagerede med “${envelope.emoji.glyph}”"
            is Envelope.Clear -> "Fjernede sin reaktion"
        }
        val label = if (showSignature) "Sendt med BetterLectio" else "#"
        return "$sentence\n\n[url=${carrierUrl(envelope)}]$label[/url]"
    }

    fun parseCarrierUrl(rawUrl: String): Envelope? {
        if (rawUrl.isBlank() || rawUrl.length > MAX_URL_LENGTH) return null
        val uri = runCatching { URI(rawUrl) }.getOrNull() ?: return null
        if (!uri.scheme.equals("https", true) || !uri.host.equals("betterlectio.dk", true)) return null
        if (uri.path.trimEnd('/') != "/download" || uri.rawQuery != null) return null
        val fragment = uri.rawFragment ?: return null
        if (!fragment.startsWith(FRAGMENT_PREFIX)) return null
        return decode(fragment.removePrefix(FRAGMENT_PREFIX))
    }

    fun parseCarrierHtml(html: String): Envelope? {
        if (html.isBlank()) return null
        val root = Jsoup.parseBodyFragment(html).body()
        val matches = root.select("a[href]").mapNotNull { anchor ->
            parseCarrierUrl(anchor.attr("href"))?.let { anchor to it }
        }
        if (matches.size != 1) return null
        val (anchor, envelope) = matches.single()
        if (anchor.text().trim() !in setOf("Sendt med BetterLectio", "#")) return null
        val clone = root.clone()
        val anchorIndex = root.select("a").indexOf(anchor)
        clone.select("a").getOrNull(anchorIndex)?.remove()
        val visibleText = MessageEditAudit.stripTerminalAudit(normalizeText(clone))
        val expected = when (envelope) {
            is Envelope.Set -> "Reagerede med “${envelope.emoji.glyph}”"
            is Envelope.Clear -> "Fjernede sin reaktion"
        }
        return envelope.takeIf { visibleText == expected }
    }

    fun resolve(messages: List<RawMessage>): ResolvedThread {
        val candidates = messages.mapIndexed { index, raw ->
            if (raw.entry.attachments.isNotEmpty()) return@mapIndexed null
            val envelope = parseCarrierHtml(raw.rawContentHtml) ?: return@mapIndexed null
            val actorKey = senderKey(raw.entry.senderEntityId, raw.entry.senderName.orEmpty())
            if (actorKey.isBlank()) return@mapIndexed null
            Carrier(
                envelope = envelope,
                actor = MessageReactionParticipant(
                    key = actorKey,
                    name = raw.entry.senderName.orEmpty(),
                    isOwn = raw.editPostbackTarget.isNotBlank(),
                ),
                editPostbackTarget = raw.editPostbackTarget,
                index = index,
            )
        }

        val occurrences = mutableMapOf<String, Int>()
        val locators = mutableMapOf<Int, MessageLocator>()
        val targetIndexes = mutableMapOf<String, Int>()
        messages.forEachIndexed { index, raw ->
            if (candidates[index] != null) return@forEachIndexed
            val locator = deriveLocator(raw.entry, occurrences) ?: return@forEachIndexed
            locators[index] = locator
            targetIndexes[locatorKey(locator)] = index
        }

        val carriersByTarget = mutableMapOf<Int, MutableMap<String, Carrier>>()
        val hidden = mutableSetOf<Int>()
        candidates.filterNotNull().forEach { carrier ->
            val targetIndex = targetIndexes[locatorKey(carrier.envelope.target)] ?: return@forEach
            if (targetIndex >= carrier.index) return@forEach
            hidden += carrier.index
            carriersByTarget.getOrPut(targetIndex) { linkedMapOf() }[carrier.actor.key] = carrier
        }

        val ownCarriers = mutableMapOf<MessageLocator, Carrier>()
        val result = messages.mapIndexedNotNull { index, raw ->
            if (index in hidden) return@mapIndexedNotNull null
            val locator = locators[index]
            val groups = linkedMapOf<MessageReactionEmoji, MutableList<MessageReactionParticipant>>()
            var ownReaction: MessageReactionEmoji? = null
            carriersByTarget[index]?.values?.forEach { carrier ->
                if (carrier.actor.isOwn && locator != null) ownCarriers[locator] = carrier
                when (val envelope = carrier.envelope) {
                    is Envelope.Clear -> Unit
                    is Envelope.Set -> {
                        groups.getOrPut(envelope.emoji) { mutableListOf() } += carrier.actor
                        if (carrier.actor.isOwn) ownReaction = envelope.emoji
                    }
                }
            }
            raw.entry.copy(
                locator = locator,
                reactions = MessageReactionEmoji.entries.mapNotNull { emoji ->
                    groups[emoji]?.let { MessageReactionGroup(emoji, it) }
                },
                ownReaction = ownReaction,
            )
        }
        return ResolvedThread(result, ownCarriers, hidden.size)
    }

    fun senderKey(entityId: String?, senderName: String): String {
        val id = entityId.orEmpty().trim()
        if (id.isNotEmpty()) return "id:$id"
        val name = Normalizer.normalize(senderName, Normalizer.Form.NFC)
            .replace(Regex("\\s+"), " ")
            .trim()
            .lowercase(danishLocale)
        return if (name.isBlank()) "" else "name:$name"
    }

    fun locatorKey(locator: MessageLocator): String =
        "${locator.senderKey}\u001f${locator.sentAt}\u001f${locator.occurrence}"

    private fun deriveLocator(
        entry: ThreadEntry,
        occurrences: MutableMap<String, Int>,
    ): MessageLocator? {
        val senderKey = senderKey(entry.senderEntityId, entry.senderName.orEmpty())
        val sentAt = entry.sentAt?.format(isoSeconds) ?: return null
        if (senderKey.isBlank()) return null
        val base = "$senderKey\u001f$sentAt"
        val occurrence = occurrences[base] ?: 0
        occurrences[base] = occurrence + 1
        return MessageLocator(senderKey, sentAt, occurrence)
    }

    private fun isValidLocator(locator: MessageLocator): Boolean =
        locator.senderKey.isNotBlank() &&
            locator.senderKey.length <= 256 &&
            locator.sentAt.matches(Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}$")) &&
            locator.occurrence in 0..9

    private fun normalizeText(root: Element): String = root.text().replace(Regex("\\s+"), " ").trim()

}

File: references/betterlectio/ios/BetterLectio/MessageReactionProtocol.swift
import Foundation
import SwiftSoup

enum MessageReactionEmoji: String, CaseIterable, Codable, Sendable {
    case thumbsUp = "👍"
    case heart = "❤️"
    case laugh = "😂"
    case surprised = "😮"
    case sad = "😢"
    case thumbsDown = "👎"
}

struct MessageLocator: Codable, Equatable, Hashable, Sendable {
    let senderKey: String
    let sentAt: String
    let occurrence: Int
}

struct MessageReactionParticipant: Codable, Equatable, Sendable {
    let key: String
    let name: String
    let isOwn: Bool
}

struct MessageReactionGroup: Codable, Equatable, Identifiable, Sendable {
    let emoji: MessageReactionEmoji
    let reactors: [MessageReactionParticipant]
    var id: MessageReactionEmoji { emoji }
}

enum MessageReactionEnvelope: Equatable, Sendable {
    case set(emoji: MessageReactionEmoji, target: MessageLocator)
    case clear(target: MessageLocator)

    var target: MessageLocator {
        switch self {
        case .set(_, let target), .clear(let target): target
        }
    }
}

enum MessageReactionProtocol {
    static let downloadURL = "https://betterlectio.dk/download"
    static let fragmentPrefix = "blr1."
    private static let maxURLLength = 2_048

    struct RawMessage: Sendable {
        let message: Message
        let rawContentHTML: String
        let editPostbackTarget: String
    }

    struct Carrier: Sendable {
        let envelope: MessageReactionEnvelope
        let actor: MessageReactionParticipant
        let editPostbackTarget: String
        let index: Int
    }

    struct ResolvedThread: Sendable {
        let messages: [Message]
        let ownCarriersByTarget: [MessageLocator: Carrier]
        let hiddenCarrierCount: Int
    }

    static func encode(_ envelope: MessageReactionEnvelope) -> String {
        let target = envelope.target
        var object: [String: Any] = [
            "v": 1,
            "target": [
                "senderKey": target.senderKey,
                "sentAt": target.sentAt,
                "occurrence": target.occurrence
            ]
        ]
        switch envelope {
        case .set(let emoji, _):
            object["op"] = "set"
            object["emoji"] = emoji.rawValue
        case .clear:
            object["op"] = "clear"
            object["emoji"] = NSNull()
        }
        guard let data = try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys]) else {
            return ""
        }
        return data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    static func decode(_ encoded: String) -> MessageReactionEnvelope? {
        guard encoded.range(of: #"^[A-Za-z0-9_-]+$"#, options: .regularExpression) != nil else { return nil }
        var base64 = encoded.replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        base64 += String(repeating: "=", count: (4 - base64.count % 4) % 4)
        guard let data = Data(base64Encoded: base64),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              object["v"] as? Int == 1,
              let targetObject = object["target"] as? [String: Any],
              let senderKey = targetObject["senderKey"] as? String,
              let sentAt = targetObject["sentAt"] as? String,
              let occurrence = targetObject["occurrence"] as? Int else { return nil }
        let target = MessageLocator(senderKey: senderKey, sentAt: sentAt, occurrence: occurrence)
        guard valid(target) else { return nil }
        switch object["op"] as? String {
        case "set":
            guard let glyph = object["emoji"] as? String,
                  let emoji = MessageReactionEmoji(rawValue: glyph) else { return nil }
            return .set(emoji: emoji, target: target)
        case "clear":
            guard object["emoji"] is NSNull else { return nil }
            return .clear(target: target)
        default:
            return nil
        }
    }

    static func carrierURL(for envelope: MessageReactionEnvelope) -> String {
        "\(downloadURL)#\(fragmentPrefix)\(encode(envelope))"
    }

    static func carrierBody(for envelope: MessageReactionEnvelope, showSignature: Bool) -> String {
        let sentence: String
        switch envelope {
        case .set(let emoji, _): sentence = "Reagerede med “\(emoji.rawValue)”"
        case .clear: sentence = "Fjernede sin reaktion"
        }
        let label = showSignature ? "Sendt med BetterLectio" : "#"
        return "\(sentence)\n\n[url=\(carrierURL(for: envelope))]\(label)[/url]"
    }

    static func parseCarrierURL(_ rawURL: String) -> MessageReactionEnvelope? {
        guard !rawURL.isEmpty, rawURL.count <= maxURLLength,
              let components = URLComponents(string: rawURL),
              components.scheme?.lowercased() == "https",
              components.host?.lowercased() == "betterlectio.dk",
              components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")) == "download",
              components.query == nil,
              let fragment = components.fragment,
              fragment.hasPrefix(fragmentPrefix) else { return nil }
        return decode(String(fragment.dropFirst(fragmentPrefix.count)))
    }

    static func parseCarrierHTML(_ html: String) -> MessageReactionEnvelope? {
        guard !html.isEmpty,
              let document = try? SwiftSoup.parseBodyFragment(html),
              let body = document.body(),
              let anchors = try? body.select("a[href]").array() else { return nil }
        let matches = anchors.compactMap { anchor -> (Element, MessageReactionEnvelope)? in
            guard let href = try? anchor.attr("href"), let envelope = parseCarrierURL(href) else { return nil }
            return (anchor, envelope)
        }
        guard matches.count == 1 else { return nil }
        let (anchor, envelope) = matches[0]
        let label = ((try? anchor.text()) ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard label == "Sendt med BetterLectio" || label == "#" else { return nil }
        let expected: String
        switch envelope {
        case .set(let emoji, _): expected = "Reagerede med “\(emoji.rawValue)”"
        case .clear: expected = "Fjernede sin reaktion"
        }
        let fullText = MessageEditAudit.strippingTerminalAudit(from: ((try? body.text()) ?? ""))
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return fullText == "\(expected) \(label)" ? envelope : nil
    }

    static func resolve(_ rawMessages: [RawMessage]) -> ResolvedThread {
        let candidates: [Carrier?] = rawMessages.enumerated().map { index, raw in
            guard raw.message.attachments.isEmpty,
                  let envelope = parseCarrierHTML(raw.rawContentHTML) else { return nil }
            let key = senderKey(entityID: raw.message.senderEntityID, senderName: raw.message.senderName)
            guard !key.isEmpty else { return nil }
            return Carrier(
                envelope: envelope,
                actor: MessageReactionParticipant(
                    key: key,
                    name: raw.message.senderName,
                    isOwn: !raw.editPostbackTarget.isEmpty
                ),
                editPostbackTarget: raw.editPostbackTarget,
                index: index
            )
        }

        var occurrences: [String: Int] = [:]
        var locators: [Int: MessageLocator] = [:]
        var targetIndexes: [String: Int] = [:]
        for (index, raw) in rawMessages.enumerated() where candidates[index] == nil {
            guard let locator = deriveLocator(raw.message, occurrences: &occurrences) else { continue }
            locators[index] = locator
            targetIndexes[locatorKey(locator)] = index
        }

        var carriersByTarget: [Int: [String: Carrier]] = [:]
        var hidden = Set<Int>()
        for carrier in candidates.compactMap({ $0 }) {
            guard let targetIndex = targetIndexes[locatorKey(carrier.envelope.target)],
                  targetIndex < carrier.index else { continue }
            hidden.insert(carrier.index)
            carriersByTarget[targetIndex, default: [:]][carrier.actor.key] = carrier
        }

        var ownCarriers: [MessageLocator: Carrier] = [:]
        var output: [Message] = []
        for (index, raw) in rawMessages.enumerated() where !hidden.contains(index) {
            let locator = locators[index]
            var grouped: [MessageReactionEmoji: [MessageReactionParticipant]] = [:]
            var ownReaction: MessageReactionEmoji?
            for carrier in (carriersByTarget[index] ?? [:]).values {
                if carrier.actor.isOwn, let locator { ownCarriers[locator] = carrier }
                if case .set(let emoji, _) = carrier.envelope {
                    grouped[emoji, default: []].append(carrier.actor)
                    if carrier.actor.isOwn { ownReaction = emoji }
                }
            }
            output.append(raw.message.with(
                locator: locator,
                reactions: MessageReactionEmoji.allCases.compactMap { emoji in
                    grouped[emoji].map { MessageReactionGroup(emoji: emoji, reactors: $0) }
                },
                ownReaction: ownReaction
            ))
        }
        return ResolvedThread(messages: output, ownCarriersByTarget: ownCarriers, hiddenCarrierCount: hidden.count)
    }

    static func senderKey(entityID: String?, senderName: String) -> String {
        let id = entityID?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !id.isEmpty { return "id:\(id)" }
        let normalized = senderName.precomposedStringWithCanonicalMapping
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased(with: Locale(identifier: "da_DK"))
        return normalized.isEmpty ? "" : "name:\(normalized)"
    }

    static func locatorKey(_ locator: MessageLocator) -> String {
        "\(locator.senderKey)\u{001F}\(locator.sentAt)\u{001F}\(locator.occurrence)"
    }

    private static func deriveLocator(_ message: Message, occurrences: inout [String: Int]) -> MessageLocator? {
        let key = senderKey(entityID: message.senderEntityID, senderName: message.senderName)
        guard !key.isEmpty, let sentAt = normalizedTimestamp(message.date) else { return nil }
        let base = "\(key)\u{001F}\(sentAt)"
        let occurrence = occurrences[base, default: 0]
        occurrences[base] = occurrence + 1
        return MessageLocator(senderKey: key, sentAt: sentAt, occurrence: occurrence)
    }

    static func normalizedTimestamp(_ value: String) -> String? {
        let regex = try? NSRegularExpression(pattern: #"(\d{1,2})-(\d{1,2})-(\d{4})\s+(\d{1,2}):(\d{2})(?::(\d{2}))?"#)
        let range = NSRange(value.startIndex..., in: value)
        guard let match = regex?.firstMatch(in: value, range: range) else { return nil }
        func capture(_ index: Int) -> String? {
            guard let range = Range(match.range(at: index), in: value) else { return nil }
            return String(value[range])
        }
        guard let day = capture(1), let month = capture(2), let year = capture(3),
              let hour = capture(4), let minute = capture(5) else { return nil }
        let second = capture(6) ?? "00"
        return "\(year)-\(month.leftPadded)-\(day.leftPadded)T\(hour.leftPadded):\(minute):\(second)"
    }

    private static func valid(_ locator: MessageLocator) -> Bool {
        !locator.senderKey.isEmpty && locator.senderKey.count <= 256 &&
            locator.sentAt.range(of: #"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}$"#, options: .regularExpression) != nil &&
            (0..<10).contains(locator.occurrence)
    }
}

private extension String {
    var leftPadded: String { count >= 2 ? self : "0" + self }
}
